src/MiniProfiler.EF6/MiniProfilerEF6.cs

```diff
@@ -10,15 +13,23 @@ namespace StackExchange.Profiling.EntityFramework6
     /// </summary>
     public static class MiniProfilerEF6
     {
+        private static readonly ConcurrentDictionary<object, object>  _resolvedDependenciesCache = new ConcurrentDictionary<object, object>();
```

Comment:
> Are we sure this doesn't grow unbounded?

Fairly confident. In the case of `IDbProviderFactoryResolver` and `IDbConnectionFactory`, its treated as a singleton and the "key" is always null. For `DbProviderServices` and `DbProviderFactory`, the key is always the provider name.

> I'd argue for 4 properly typed dictionaries here

I like that, very clean. Wish I thought of it originally.

This style of caching actually mimics the caching that EF does (there's an internal `CachingDependencyResolver` that they use for the core resolvers of these same types which behave the exact same way).

This shouldn't change the behavior if something else replaces the service either before or after we do - however there's probably a good chance that if something else were replacing these services, things won't work well... I'm not sure we can totally solve that.